### DISCUSSION TITLE
Execute and set Title

### DISCUSSION THREAD

**Discussion by Simon-Eklundh** [i3/i3#4556] [Q&A]
Tl;Dr: how do I format a title in a program on startup without needing for_window.

Hi! I recently decided to create two discord accounts I want to have open at the same time.  I decided to solve this using discord and discord-canary.

However, these two have the same class, so I wanted to change their class/title. To make automatic layouts possible.

I've tried a few different ways found in a GitHub post from two years ago, primarily, but in my case I've tried to make it an auto start with a Mark or setting title immediately:

exec discord , mark personal
I've also tried
exec discord , exec "sleep 3 && i3-msg Mark personal"

Neither of those marked my discord windows. I've also tried the equivalent with instant title;

exec discord, title_format "personal %title"

As I'm now stuck I'd like to see if someone if you can give me a hand.

Sincerely Simon

Edit: I solved it by making a bash script that starts the first program, focuses it, runs an action on it, then starts the second program.

**Comment by Airblader** [comment]


**Reply by Simon-Eklundh** [reply]
So it's not possible to execute and modify title, I kind of expected that.

However, how about marking? I know it has worked previously from this thread: https://www.reddit.com/r/i3wm/comments/bg1qd9/keybinding_execute_and_mark_at_the_same_time/?utm_medium=android_app&utm_source=share

Although it might be the application allowing it because of some reason...

**Reply by Airblader** [reply]
Marking a window requires you to know which window to mark. It's a catch-22.

**Reply by Airblader** [reply]
(Sorry, posted my comment multiple times due to flaky connection and deleted this one again)

**Reply by Simon-Eklundh** [reply]
> (Sorry, posted my comment multiple times due to flaky connection and deleted this one again)

All good, don't worry!

**Reply by Simon-Eklundh** [reply]
> Marking a window requires you to know which window to mark.

I assume you haven't got some sort of way of getting around it?

**Reply by Airblader** [reply]
Only what I mentioned in the edit to my original comment. Apart from that the best you can probably do is solve things via timing, which is slow and unreliable.

**Reply by Simon-Eklundh** [reply]
> Only what I mentioned in the edit to my original comment. Apart from that the best you can probably do is solve things via timing, which is slow and unreliable.

Alright, thanks!

If you're relayed to the project: I remember seeing someone propose something similar to

ExecAndMark program

On GitHub a while back (think they solved it by using a -t flag on a terminal so they closed the ticket)

**Reply by Airblader** [reply]
If you can tell me how that is supposed to work we can talk about it for sure. As I outlined above, a process and a window are very different concepts.

**Reply by Simon-Eklundh** [reply]
Sorry for late reply, life (and not getting push notifications) got in the way


If we think of execute&mark (as that can then be used for the rest)

Then I'm thinking syntax:

ExecAndMark program nameOfMark

And actual execution is a bit more difficult. Theoretically you could either keep focus on the window getting opened until it is marked, this should work except for programs that spawn one window that is replaced (I'm not sure if that's how discord does it, but it might be. Same possibility with the jetbrains suite).
Or
It exec:s the program and then does a for_window for the class (or window Id if that's accessible for you)



I got around it by writing a bash script that started one program at a time, marking the first one and then starting the second one (kinda, I didn't actually mark, because I wanted something different)

**Reply by Airblader** [reply]
I don't see you address the problem that we have. If you execute `run_something`, how do you know whether it will ever open a window, when it will open it, or, when some window opens, that it comes from this process? This is the crux of this issue: we only know about windows, not processes, and cannot correlate them.

**Reply by Simon-Eklundh** [reply]
that is a fair point.

I can't really think of a solution then.

oh well, cheers!

**Comment by Airblader** [comment]
> how do I format a title in a program on startup without needing for_window.

Unless the application supports a command line flag to set the title, you can't (reliably).

Executing a binary and a window being mapped are different things, the window manager only knows about the latter, and you don't know whether a process will create any windows, how many it will create, or when it will create them.

If Discord sets _NET_WM_PID on its windows you could look up the process based on that and potentially differentiate them in some way. But that's very hypothetical and outside of i3.
